# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [1]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

os.getcwd()

'C:\\Users\\Billy\\Desktop\\Springboard\\Guided Capstone'

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [2]:
df = pd.read_csv('..\Guided Capstone\data\step3_output.csv')
df.drop('Unnamed: 0', axis='columns', inplace=True)
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1               Arizona Snowbowl     Arizona        11500           2300   
2            Sunrise Park Resort     Arizona        11100           1800   
3  Yosemite Ski & Snowboard Area  California         7800            600   
4                  Bear Mountain  California         8805           1665   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          1          0     2       2  ...   
2      0        0.0          0          1     2       3  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000             150.0       36.0   
1              777.0      104.000000             122.0       81.0   
2              800.0       80.000000             115.0       49.0   
3               88.0      174.873239             110.0       84.0   
4              198.0      198.000000             122.0       76.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            260.0     89.000000      89.00000              122.0   
2            250.0     74.000000      78.00000              104.0   
3            300.0     47.000000      47.00000              107.0   
4            100.0     57.916957      64.16681              130.0   

   NightSkiing_ac  cluster  
0            30.0        0  
1             0.0        1  
2            80.0        1  
3             0.0        1  
4             0.0        1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [3]:
df1 = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
df1.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1               Arizona Snowbowl        11500           2300      0   
2            Sunrise Park Resort        11100           1800      0   
3  Yosemite Ski & Snowboard Area         7800            600      0   
4                  Bear Mountain         8805           1665      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          1          0     2       2       1  ...             0   
2        0.0          0          1     2       3       1  ...             0   
3        0.0          0          0     0       1       3  ...             0   
4        0.0          0          2     1       2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

In [4]:
df1.shape

(243, 60)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [5]:
from sklearn import preprocessing

In [6]:
df1 = df1.drop(['Name', 'AdultWeekend'], axis=1)

In [7]:
stand = preprocessing.scale(df1)

In [8]:
y = df['AdultWeekend']

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [9]:
from sklearn.model_selection import train_test_split

In [10]:
x_train, x_test, y_train, y_test = train_test_split(stand, y, test_size=0.25)

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(x_train, y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [12]:
y_pred = lm.predict(x_test)

In [13]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

-8.100736007221314e+21
206798333844.44266


## Review Model Outcomes — Iterate over additional models as needed.

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [14]:
model.intercept_

-51912338886.74111

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [15]:
coef = pd.Series(abs(model.coef_))

In [16]:
variables = pd.Series(df1.columns)

In [17]:
pd.concat([variables, coef], axis=1).sort_values(1, ascending=False)

0             1
3           fastEight  7.519447e+13
10       total_chairs  5.248961e+13
8              double  2.795990e+13
9             surface  2.612142e+13
7              triple  2.318803e+13
6                quad  1.869056e+13
5           fastQuads  1.420193e+13
4           fastSixes  5.102864e+12
43           New York  4.027809e+12
35           Michigan  3.763555e+12
2               trams  3.699080e+12
47       Pennsylvania  3.303681e+12
56          Wisconsin  3.035873e+12
25         California  3.035873e+12
40      New Hampshire  2.939357e+12
36          Minnesota  2.838613e+12
52            Vermont  2.505812e+12
34      Massachusetts  2.382318e+12
26           Colorado  2.250865e+12
42         New Mexico  2.109967e+12
51               Utah  1.957584e+12
28              Idaho  1.957584e+12
44     North Carolina  1.790788e+12
32              Maine  1.790788e+12
27        Connecticut  1.790788e+12
54         Washington  1.790788e+12
46             Oregon  1.790788e+12
57            Wyoming  1.790788e+12
45               Ohio  1.790788e+12
53           Virginia  1.605091e+12
55      West Virginia  1.605091e+12
29           Illinois  1.605091e+12
38            Montana  1.605091e+12
31               Iowa  1.392954e+12
41         New Jersey  1.139709e+12
39             Nevada  1.139709e+12
49       South Dakota  1.139709e+12
24            Arizona  1.139709e+12
37           Missouri  1.139709e+12
30            Indiana  1.139709e+12
50          Tennessee  8.075665e+11
33           Maryland  8.075665e+11
23             Alaska  8.075665e+11
19       AdultWeekday  1.168909e+01
48       Rhode Island  3.211418e+00
20  projectedDaysOpen  1.760742e+00
11               Runs  1.564087e+00
1       vertical_drop  1.483749e+00
18    averageSnowfall  1.332031e+00
14  SkiableTerrain_ac  1.298828e+00
15     Snow Making_ac  8.789062e-01
16   daysOpenLastYear  7.792969e-01
12       TerrainParks  6.445312e-01
13      LongestRun_mi  4.921875e-01
22            cluster  4.233398e-01
21     NightSkiing_ac  2.988281e-01
0         summit_elev  1.902468e-01
17          yearsOpen  3.906250e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [18]:
#select original df and drop Name, AdultWeekend, and state columns
df2 = df.drop(['Name', 'AdultWeekend', 'state'], 1)
df2.head()

summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0         2090            294      0        0.0          0          0     0   
1        11500           2300      0        0.0          1          0     2   
2        11100           1800      0        0.0          0          1     2   
3         7800            600      0        0.0          0          0     0   
4         8805           1665      0        0.0          0          2     1   

   triple  double  surface  ...  LongestRun_mi  SkiableTerrain_ac  \
0       1       0        2  ...            1.0               30.0   
1       2       1        2  ...            2.0              777.0   
2       3       1        0  ...            1.2              800.0   
3       1       3        1  ...            0.4               88.0   
4       2       3        4  ...            1.5              198.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0       30.000000             150.0       36.0             69.0     30.000000   
1      104.000000             122.0       81.0            260.0     89.000000   
2       80.000000             115.0       49.0            250.0     74.000000   
3      174.873239             110.0       84.0            300.0     47.000000   
4      198.000000             122.0       76.0            100.0     57.916957   

   projectedDaysOpen  NightSkiing_ac  cluster  
0              152.0            30.0        0  
1              122.0             0.0        1  
2              104.0            80.0        1  
3              107.0             0.0        1  
4              130.0             0.0        1  

[5 rows x 23 columns]

In [19]:
#standardize the scale of the features
stand2 = preprocessing.scale(df2)

In [20]:
#split into training and testing datasets
x_train, x_test, y_train, y_test = train_test_split(stand2, y, test_size=0.25)

In [21]:
#train model
model2 = lm.fit(x_train, y_train)

In [22]:
#predict y
y_pred = lm.predict(x_test)

In [23]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.8278784968282618
6.814855025627383


In [24]:
#intercept value of the model
model2.intercept_

60.43073380579439

In [25]:
#absolute values of coefficiencies for each variable
coef2 = pd.Series(abs(model2.coef_))

In [26]:
#variable names
variables2 = pd.Series(df2.columns)

In [27]:
#coefficiencies in descending order
pd.concat([variables2, coef2], axis=1).sort_values(1, ascending=False)

0             1
19       AdultWeekday  1.317573e+01
0         summit_elev  2.459583e+00
22            cluster  2.132637e+00
16   daysOpenLastYear  2.128434e+00
20  projectedDaysOpen  1.974514e+00
6                quad  1.668378e+00
1       vertical_drop  1.419105e+00
18    averageSnowfall  1.338385e+00
2               trams  1.267389e+00
11               Runs  9.540252e-01
7              triple  9.300115e-01
9             surface  8.051680e-01
12       TerrainParks  5.914516e-01
14  SkiableTerrain_ac  5.651074e-01
10       total_chairs  5.567920e-01
13      LongestRun_mi  4.904149e-01
15     Snow Making_ac  3.707260e-01
17          yearsOpen  3.426417e-01
5           fastQuads  3.422721e-01
4           fastSixes  3.230638e-01
21     NightSkiing_ac  4.066032e-02
8              double  3.627810e-02
3           fastEight  1.110223e-16

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [28]:
#select original df and drop Name, AdultWeekend, state, and summit_elev columns
df3 = df.drop(['Name', 'AdultWeekend', 'state', 'summit_elev'], 1)
df3.head()

vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  triple  \
0            294      0        0.0          0          0     0       1   
1           2300      0        0.0          1          0     2       2   
2           1800      0        0.0          0          1     2       3   
3            600      0        0.0          0          0     0       1   
4           1665      0        0.0          0          2     1       2   

   double  surface  total_chairs  ...  LongestRun_mi  SkiableTerrain_ac  \
0       0        2             3  ...            1.0               30.0   
1       1        2             8  ...            2.0              777.0   
2       1        0             7  ...            1.2              800.0   
3       3        1             5  ...            0.4               88.0   
4       3        4            12  ...            1.5              198.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0       30.000000             150.0       36.0             69.0     30.000000   
1      104.000000             122.0       81.0            260.0     89.000000   
2       80.000000             115.0       49.0            250.0     74.000000   
3      174.873239             110.0       84.0            300.0     47.000000   
4      198.000000             122.0       76.0            100.0     57.916957   

   projectedDaysOpen  NightSkiing_ac  cluster  
0              152.0            30.0        0  
1              122.0             0.0        1  
2              104.0            80.0        1  
3              107.0             0.0        1  
4              130.0             0.0        1  

[5 rows x 22 columns]

In [29]:
#standardize the scales of the features
stand3 = preprocessing.scale(df3)

In [30]:
#split into training and testing datasets
x_train, x_test, y_train, y_test = train_test_split(stand3, y, test_size=0.25)

In [31]:
#train model
model3 = lm.fit(x_train, y_train)

In [32]:
#predict y
y_pred = lm.predict(x_test)

In [33]:
print(explained_variance_score(y_test, y_pred))
print(mean_absolute_error(y_test, y_pred))

0.7690884954012668
5.935931999630803


In [34]:
#model intercept
model3.intercept_

61.1028250063538

In [35]:
#absolute values of coefficiencies for each variable convert to series
coef3 = pd.Series(abs(model3.coef_))

In [36]:
#variable names convert to series
variables3 = pd.Series(df3.columns)

In [37]:
#coefficiencies sorted in descending order
pd.concat([variables3, coef3], axis=1).sort_values(1, ascending=False)

0             1
18       AdultWeekday  1.263084e+01
13  SkiableTerrain_ac  1.846672e+00
5                quad  1.828528e+00
15   daysOpenLastYear  1.615861e+00
3           fastSixes  1.577684e+00
10               Runs  1.453271e+00
21            cluster  1.360932e+00
17    averageSnowfall  1.335391e+00
19  projectedDaysOpen  1.311780e+00
8             surface  1.197166e+00
6              triple  1.098327e+00
1               trams  1.097881e+00
4           fastQuads  9.663322e-01
0       vertical_drop  9.637503e-01
9        total_chairs  8.943681e-01
14     Snow Making_ac  8.723015e-01
12      LongestRun_mi  8.372673e-01
11       TerrainParks  7.415300e-01
16          yearsOpen  4.376182e-01
20     NightSkiing_ac  2.684399e-01
7              double  3.072645e-02
2           fastEight  3.330669e-15

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -8.100736007221314e+21 | 206798333844.44266 |-|
| Model 2. | 0.83 | 6.81 |'state'|
| Model 3. | 0.77 | 5.94 |'state','summit_elev','base_elev'|

Model Selection: 2